### PR TITLE
[Refactoring] Decouple and encapsulate tier two synchronization state

### PR DESCRIPTION
**Pull request by furszy**
Work decoupled from #2647 which has grew a lot lately. So it can get reviewed and merged in parallel of #2607.

Essentially have encapsulated the tier two synchronization status in a separate object which is only in charge of storing and retrieving the sync status members values in a synchronic manner.

Under this new architecture, every manager/object that needs access to it can add a small dependency to the new `g_tiertwo_sync_state` object only (which has no other dependencies), without depending on the whole `masternodesSync` object (which is a bad name for the tier two sync manager, but.. renaming will come in another PR), which depends on every other tier two manager and lot other files.
Plus, to not generate confusions, the `g_tiertwo_sync_state` members update are only, and exclusively, done by the tier two sync manager.

Fixing in this way, at least (because there are surely more), the following problems:

1) Circular dependencies:

   * "activemasternode -> masternode-sync -> activemasternode".
   * "budget/budgetmanager -> masternode-sync -> budget/budgetmanager"
   * "masternode -> masternode-sync -> masternode"
   * "masternode-payments -> masternode-sync -> masternode-payments"
   * "masternode-sync -> validation -> masternode-sync"
   * "evo/deterministicmns -> masternode -> masternode-sync -> evo/deterministicmns"

2) Allow and guard concurrent access. (Prior to this work, the word "concurrency" was inexistent for this code).

### CODE REVIEW

**Review by random-zebra — COMMENTED** [review]
Very nice refactoring. Code ACK with some notes.

File: src/masternode-sync.cpp [hunk 1]
```diff
@@ -50,9 +51,9 @@ bool CMasternodeSync::NotCompleted()
             sporkManager.IsSporkActive(SPORK_13_ENABLE_SUPERBLOCKS)));
 }
 
-bool CMasternodeSync::IsBlockchainSynced()
+bool CMasternodeSync::UpdateBlockchainSynced()
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/masternode-sync.cpp:38]: yeah, good.

File: src/masternode-sync.cpp [hunk 2]
```diff
@@ -50,9 +51,9 @@ bool CMasternodeSync::NotCompleted()
...
 {
-    if (fBlockchainSynced) return true;
+    if (g_tiertwo_sync_state.IsBlockchainSynced()) return true;
     if (fImporting || fReindex) return false;
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/masternode-sync.cpp]: If we reset the chain sync state here, we would introduce a race condition. Another thread could access the chain sync variable in the middle of this function.

Check e65c763. Have gone further and unified the chain sync status update in a single line + added a time delay between updates so this function does not lock, every 5 secs, a mutex that is being locked on the main processing thread as well.

**Comment by furszy** [comment]
Updated per feedback 👍 .

File: src/masternode-sync.cpp [hunk 3]
```diff
@@ -270,7 +192,19 @@ void CMasternodeSync::Process()
 
     if (tick++ % MASTERNODE_SYNC_TIMEOUT != 0) return;
 
-    if (IsSynced()) {
+    // if the last call to this function was more than 60 minutes ago (client was in sleep mode)
+    // reset the sync process
+    int64_t now = GetTime();
+    if (lastProcess != 0 && now > lastProcess + 60 * 60) {
+        Reset();
+        g_tiertwo_sync_state.SetBlockchainSync(false, 0);
```
Comment (random-zebra) [inline src/masternode-sync.cpp:200]: `Reset()` already sets the blockchain-sync to false, no need to do it again here (not introduced in this PR, but it's a good opportunity to fix it).

Comment (furszy) [reply src/masternode-sync.cpp:200]: sure

**Comment by furszy** [comment]
Updated per feedback, squashed it in your 4b3bcfb zebra.

**Review by random-zebra — DISMISSED** [review]
ACK 4b3bcfb95145eb4ee7905899daafa62cb253c7b7

**Comment by furszy** [comment]
Just noticed that need to rebase this one once more, the new test `dkg_pose` introduced in #2607 is setting the tier two sync asset manually.
So.. if we merge it as is, the final branch will not compile.

**Comment by furszy** [comment]
rebased, ready to go once more.

**Review by random-zebra — APPROVED** [review]
ACK 24d9fa5a554068a6261328b1b5e60562e9a2fa41

**Comment by furszy** [comment]
due conflicts with recently merged PRs, rebased on master once more.

**Review by random-zebra — APPROVED** [review]
ACK 75bc24f708b2f6c4d8de1ca53a462ca882b15df1

**Comment by furszy** [comment]
rebased for merge conflicts with 2697, had to dismiss the approval again.

**Review by random-zebra — APPROVED** [review]
ACK 3e05c82aadcf8954e507636e74a39a5bfecd84d5 again. Let's get this merged.

**Pull request merged**